components/support/sql/src/each_chunk.rs

```diff
@@ -0,0 +1,180 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use rusqlite::{
+    self,
+    limits::Limit,
+    types::ToSql,
+};
+
+/// Returns SQLITE_LIMIT_VARIABLE_NUMBER as read from an in-memory connection and cached.
+/// connection and cached. That means this will return the wrong value if it's set to a lower
+/// value for a connection using this will return the wrong thing, but doing so is rare enough
+/// that we explicitly don't support it (why would you want to lower this at runtime?).
+///
+/// If you call this and the actual value was set to a negative number or zero (nothing prevents
+/// this beyond a warning in the SQLite documentation), we panic. However, it's unlikely you can
+/// run useful queries if this happened anyway.
+pub fn default_max_variable_number() -> usize {
+    lazy_static! {
+        static ref MAX_VARIABLE_NUMBER: usize = {
+            let conn = rusqlite::Connection::open_in_memory()
+                .expect("Failed to initialize in-memory connection (out of memory?)");
+
+            let limit = conn.limit(Limit::SQLITE_LIMIT_VARIABLE_NUMBER);
+            assert!(limit > 0,
+                    "Illegal value for SQLITE_LIMIT_VARIABLE_NUMBER (must be > 0) {}",
+                    limit);
+            limit as usize
+        };
+    }
+    *MAX_VARIABLE_NUMBER
+}
+
+/// Helper for the case where you have a `&[impl ToSql]` of arbitrary length, but need a
+/// `&[&dyn ToSql]` of no more than the connection's `MAX_VARIABLE_NUMBER` (rather,
+/// `default_max_variable_number()`). This is useful when performing batched updates.
+///
+/// The `do_chunk` callback is called with a slice of no more than `default_max_variable_number()`
+/// items as it's first argument, and the offset from the start as it's second.
+///
+/// See `each_chunk_mapped` for the case where `T` doesn't implement `ToSql`, but converting to
+/// `&dyn ToSql` is nonetheless possible.
+pub fn each_chunk<'a, T, E, F>(items: &[T], do_chunk: F) -> Result<(), E>
+where
+    T: ToSql + 'a,
+    F: FnMut(&[&dyn ToSql], usize) -> Result<(), E>,
+{
+    each_sized_chunk_mapped(items, default_max_variable_number(), |t| t as &dyn ToSql, do_chunk)
+}
+
+/// A version of `each_chunk` for the case when the conversion to `to_sql` requires an custom
+/// intermediate step. For example, you might want to grab a property off of an arrray of records
+pub fn each_chunk_mapped<'a, T, E, Mapper, DoChunk>(
```

Comment:
We could maybe implement this on `ConnExt`, which would also let us query that specific connection's limit...but I think this makes sense, too. I also agree that we don't need to support lowering the limit at runtime. 👍